mp4/mdat.go

```diff
@@ -84,3 +87,42 @@ func (m *MdatBox) HeaderSize() uint64 {
 func (m *MdatBox) PayloadAbsoluteOffset() uint64 {
 	return m.StartPos + m.HeaderSize()
 }
+
+// AddReadSeeker - add readseeker to Mdat box.
+// When a file is decoded lazily, the Mdat Data byte slice is nil
+// and this readseeker is to read data whenever the data is needed.
+func (m *MdatBox) AddReadSeeker(rs io.ReadSeeker) {
+	m.readSeeker = rs
+}
+
+// ReadData reads Mdat data specified by the start and size.
+// Input argument start is the postion relative to the start of a file.
+func (m *MdatBox) ReadData(start, size int64) ([]byte, error) {
+	// The Mdat box was decoded lazily
+	if m.decLazyDataSize > 0 {
+		if m.readSeeker == nil {
+			return nil, errors.New("lazy mdat mode expects readseeker to read data")
+		}
+		_, err := m.readSeeker.Seek(start, io.SeekStart)
+		if err != nil {
+			return nil, fmt.Errorf("unable to seek to %d", start)
+		}
+
+		buf := make([]byte, size)
+		n, err := m.readSeeker.Read(buf)
+		if err != nil {
+			return nil, err
+		}
+		if int64(n) != size {
+			return nil, fmt.Errorf("expect to read %d bytes, but only read %d bytes", size, n)
+		}
+		return buf, nil
+	}
+
+	// All Mdat Data is in memory
+	mdatPayloadStart := m.PayloadAbsoluteOffset()
+	offsetInMdatData := uint64(start) - mdatPayloadStart
+
+	return m.Data[offsetInMdatData : offsetInMdatData+uint64(size)], nil
+
```

Comment:
I think this looks nice, but we need some test for it, and ideally integration in the example/segmenter code.

Reply:
added a couple of test cases for ReadData method. I'm currently working on an integration example.